packages/playground/src/components/repo-settings/repo-branch-settings-rules/repo-branch-settings-rules-fields.tsx

```diff
@@ -0,0 +1,287 @@
+          <FormFieldSet.Message theme={MessageTheme.ERROR}>
+            {errors!.targetPatterns.message?.toString()}
+          </FormFieldSet.Message>
+        )}
+      </div>
+    </FormFieldSet.ControlGroup>
+  )
+}
+
+export const BranchSettingsRuleDefaultBranchField: React.FC<FieldProps> = ({ register, errors, watch, setValue }) => (
+  <FormFieldSet.ControlGroup className="min-h-8">
+    <FormFieldSet.Option
+      control={
+        <Checkbox
+          {...register!('defaultBranchValue')}
+          checked={watch!('defaultBranchValue')}
+          onCheckedChange={() => setValue!('defaultBranchValue', !watch!('defaultBranchValue'))}
+          id="default-branch"
+        />
+      }
+      id="default-branch"
+      label="Default Branch"
+    />
+
+    {errors!.defaultBranchValue && (
+      <FormFieldSet.Message theme={FormFieldSet.MessageTheme.ERROR}>
+        {errors!.defaultBranchValue.message?.toString()}
+      </FormFieldSet.Message>
+    )}
+  </FormFieldSet.ControlGroup>
+)
+
+export const BranchSettingsRuleBypassListField: React.FC<FieldProps> = ({ watch, setValue }) => (
+  <FormFieldSet.ControlGroup>
+    <FormFieldSet.Label htmlFor="bypassValue">Bypass list</FormFieldSet.Label>
+    <Select
+      value={watch!('bypassValue')}
+      onValueChange={value => setValue!('bypassValue', value, { shouldValidate: true })}>
+      <SelectTrigger id="bypassValue">
+        <SelectValue placeholder="Select users" />
+      </SelectTrigger>
+      <SelectContent>
+        <SelectItem value="1">user 1</SelectItem>
+        <SelectItem value="2">user 2</SelectItem>
+        <SelectItem value="3">user 3</SelectItem>
+      </SelectContent>
+    </Select>
+  </FormFieldSet.ControlGroup>
+)
+
+export const BranchSettingsRuleEditPermissionsField: React.FC<FieldProps> = ({ register, errors, watch, setValue }) => (
+  <FormFieldSet.ControlGroup className="min-h-8">
+    <FormFieldSet.Option
+      control={
+        <Checkbox
+          {...register!('editPermissionsValue')}
+          checked={watch!('editPermissionsValue')}
+          onCheckedChange={() => setValue!('editPermissionsValue', !watch!('editPermissionsValue'))}
+          id="edit-permissons"
+        />
+      }
+      id="edit-permissons"
+      label="Allow users with edit permission on the repository to bypass"
+    />
+
+    {errors!.editPermissionsValue && (
+      <FormFieldSet.Message theme={FormFieldSet.MessageTheme.ERROR}>
+        {errors!.editPermissionsValue.message?.toString()}
+      </FormFieldSet.Message>
+    )}
+  </FormFieldSet.ControlGroup>
+)
+
+export const BranchSettingsRuleListField: React.FC<FieldProps> = ({ watch, setValue }) => {
+  const handleCheckboxChange = (ruleId: string, checked: boolean) => {
+    const updatedRules = watch!('rules').map(r => (r.id === ruleId ? { ...r, checked } : r))
+    setValue!('rules', updatedRules)
+  }
+  const handleSubmenuChange = (ruleId: string, submenuId: string, checked: boolean) => {
```

Comment:
not a fan of manually maintaining form state in the components. few alternatives to consider:

1. use `useReducer` and move it out of the component
2. let the form store the raw data, and do this processing once when the form is being submitted